### PR TITLE
Add flag to warn about unreachable branches and exprs

### PR DESCRIPTION
**Pull request by Michael0x2a**
This diff adds a `--warn-unreachable` flag that reports an error if:

1. Any statement is inferred to be unreachable
2. Any subexpression in boolean expressions, inline if statements, and list/set/generator/dict comprehensions are always unreachable or redundant.

This only takes into account the results of *type* analysis. We do not report errors for statements that are inferred to be unreachable in the semantic analysis phase (e.g. due to `sys.platform` checks and the like).

Those checks are all intentional/deliberately flag certain branches as unreachable, so error messages would be annoying in those cases.

A few additional notes:

1. I didn't spend a huge amount of time trying to come up with error messages. They could probably do with some polishing/rewording.

2. I thought about enabling this flag for default with mypy, but unfortunately that ended up producing ~40 to 50 (apparently legitimate) error messages.

   About a fourth of them were due to runtime checks checking to see if some type is None (despite not being declared to be Optional), about a fourth seemed to be due to mypy not quite understanding how to handle things like traits and Bogus[...], and about a fourth seemed to be legitimately unnecessary checks we could safely remove.

   The final fourth was a mixture of typeshed bugs and misc errors with the reachability checks. (e.g. see https://github.com/python/mypy/pull/7048)

3. For these reasons, I decided against adding this flag to `--strict` and against documenting it yet. We'll probably need to flush out a few bugs in mypy first/get mypy to the state where we can at least honestly dogfood this.

Resolves https://github.com/python/mypy/issues/2395; partially addresses https://github.com/python/mypy/issues/7008.

### CODE REVIEW

File: mypy/main.py [hunk 1]
```diff
@@ -492,6 +492,10 @@ def add_invertible_flag(flag: str,
                         help="Treat imports as private unless aliased",
                         group=strictness_group)
 
+    add_invertible_flag('--disallow-inferred-unreachable', default=False, strict_flag=False,
```
Comment (JukkaL) [inline mypy/main.py]: Maybe we can make the flag name a bit clearer. What about `--warn-unreachable`? We still allow some unreachable code, but I think that we don't need to make this distinction in the flag name. I think that it's sufficient to describe the semantics in the documentation.

File: test-data/unit/check-unreachable-code.test [hunk 2]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
...
+    reveal_type(x)
+[builtins fixtures/exception.pyi]
+
+[case testUnreachableFlagExpressions]
+# flags: --disallow-inferred-unreachable
+def foo() -> bool: ...
+
+lst = [1, 2, 3, 4]
+
+a = True or foo()                        # E: The right operand of this 'or' expression is never evaluated
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: Idea for message: `Right operand of 'or' is never evaluated`

File: test-data/unit/check-unreachable-code.test [hunk 3]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
...
+b = False or foo()                       # E: The left operand of this 'or' expression is always false
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: Similar to above, maybe rephrase as `Left operand of 'or' is always false`

File: test-data/unit/check-unreachable-code.test [hunk 4]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
...
+c = True and foo()                       # E: The left operand of this 'and' expression is always true
+d = False and foo()                      # E: The right operand of this 'and' expression is never evaluated
+e = True or (True or (True or foo()))    # E: The right operand of this 'or' expression is never evaluated
+f = (True or foo()) or (True or foo())   # E: The right operand of this 'or' expression is never evaluated
+g = 3 if True else 4                     # E: The 'else' expression in this inline if expression is never evaluated
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: Idea for message: `If condition is always true`

File: test-data/unit/check-unreachable-code.test [hunk 5]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
...
+h = 3 if False else 4                    # E: The 'if' expression in this inline if expression is never evaluated
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: Idea for message: `If condition is always false`

File: test-data/unit/check-unreachable-code.test [hunk 6]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
...
+i = [x for x in lst if True]             # E: The conditional check in this comprehension is always true
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: Idea for message: `If condition in comprehension is always true`

File: test-data/unit/check-unreachable-code.test [hunk 7]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
...
+if False:
+    assert False
+    reveal_type(x)
+
+if False:
+    raise Exception()
+    reveal_type(x)
+
+if False:
+    assert_never(x)
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: Random thought (no need to address in this PR): What do you think about providing a magic utility function such as `unreachable()` for easily silencing errors about unreachable code?

File: test-data/unit/check-unreachable-code.test [hunk 8]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
...
+[builtins fixtures/exception.pyi]
+
+[case testUnreachableFlagIgnoresSemanticAnalysisUnreachable]
+# flags: --disallow-inferred-unreachable --python-version 3.7 --platform win32 --always-false FOOBAR
+import sys
+from typing import TYPE_CHECKING
+
+x: int
+if TYPE_CHECKING:
+    reveal_type(x)  # N: Revealed type is 'builtins.int'
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: Also test that any code is allowed in an else block?

File: test-data/unit/check-unreachable-code.test [hunk 9]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
...
+
+if sys.platform == 'darwin':
+    reveal_type(x)
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: Similar to above, test that any code is allowed in an else block if the platform check is true?

File: test-data/unit/check-unreachable-code.test [hunk 10]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
...
+
+if sys.version_info == (2, 7):
+    reveal_type(x)
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: Similar to above, test that any code is allowed in an else block if the version info check is true?

File: test-data/unit/check-unreachable-code.test [hunk 11]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
...
 reveal_type('')  # N: Revealed type is 'builtins.str'
 [builtins fixtures/ops.pyi]
+
+[case testUnreachableFlagWithBadControlFlow]
+# flags: --disallow-inferred-unreachable
+a: int
+if isinstance(a, int):
+    reveal_type(a)  # N: Revealed type is 'builtins.int'
+else:
+    reveal_type(a)  # E: This branch is inferred to be unreachable
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: Idea for message: `Statement is unreachable`

Even better, maybe we can somehow detect that this is caused by an always-true `isinstance` check and generate an error on the line with the `isinstance` check.

File: test-data/unit/check-unreachable-code.test [hunk 12]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
     reveal_type('')  # N: Revealed type is 'builtins.str'
 reveal_type('')  # N: Revealed type is 'builtins.str'
 [builtins fixtures/ops.pyi]
+
+[case testUnreachableFlagWithBadControlFlow]
+# flags: --disallow-inferred-unreachable
+a: int
+if isinstance(a, int):
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: Add test with a function that accepts a type variable with value restriction and performs an `isinstance` check that is always true/false for some value of the type variable. It may be reasonable to disable unreachability checks in a function with a type variable restriction (at least for now), since the correct implementation may be tricky and such type variables are relatively rare.

File: mypy/checker.py [hunk 13]
```diff
@@ -1777,13 +1777,36 @@ def check_import(self, node: ImportBase) -> None:
...
             self.binder.unreachable()
             return
         for s in b.body:
             if self.binder.is_unreachable():
+                if self.options.disallow_inferred_unreachable and not self.is_raising_or_empty(s):
+                    self.msg.unreachable_branch(s)
                 break
             self.accept(s)
 
+    def is_raising_or_empty(self, s: Statement) -> bool:
```
Comment (JukkaL) [inline mypy/checker.py]: Add docstring which describes the purpose of this function.

File: mypy/messages.py [hunk 14]
```diff
@@ -1228,6 +1228,30 @@ def note_call(self, subtype: Type, call: Type, context: Context) -> None:
         self.note('"{}.__call__" has type {}'.format(self.format_bare(subtype),
                                                      self.format(call, verbosity=1)), context)
 
+    def unreachable_branch(self, context: Context) -> None:
```
Comment (JukkaL) [inline mypy/messages.py]: What do you think about renaming this to `unreachable_statement`? Would it be clearer (and correct)?

File: mypy/messages.py [hunk 15]
```diff
@@ -1228,6 +1228,30 @@ def note_call(self, subtype: Type, call: Type, context: Context) -> None:
...
+        self.fail("This branch is inferred to be unreachable", context)
+
+    def always_same_truth_value_left_operand(self, op_name: str, context: Context) -> None:
```
Comment (JukkaL) [inline mypy/messages.py]: Document supported op names.

File: test-data/unit/check-unreachable-code.test [hunk 16]
```diff
@@ -715,3 +715,179 @@ if sys.version_info[0] >= 2:
     reveal_type('')  # N: Revealed type is 'builtins.str'
 reveal_type('')  # N: Revealed type is 'builtins.str'
 [builtins fixtures/ops.pyi]
+
+[case testUnreachableFlagWithBadControlFlow]
```
Comment (JukkaL) [inline test-data/unit/check-unreachable-code.test]: What about something like `DEBUG: Final = False` followed by `if DEBUG: print('foo')`? Is there a test for this use case?

Comment (Michael0x2a) [reply test-data/unit/check-unreachable-code.test]: Our reachability checks currently don't special-case Literal bools (or Finals that are bools) at all -- we'd actually just end up checking both branches in the `if DEBUG: ... else: ...` block.

I'm not sure exactly what behavior we want though, which is why I refrained from adding a test case.

(On one hand, it'd be nice to make mypy smarter about reachability in general; on the other improving mypy's understanding would lead to actual regressions in type inference if users don't opt-in to using this new flag.)

File: mypy/messages.py (hunk 14 above)
Comment (Michael0x2a) [reply mypy/messages.py]: Good idea! I think it'd be more correct, especially since we're also reporting these errors after things like asserts and returns. (I also added a test case for the latter).

File: test-data/unit/check-unreachable-code.test (hunk 11 above)
Comment (Michael0x2a) [reply test-data/unit/check-unreachable-code.test]: I experimented a bit with this idea. I think switching to reporting an error on the check might be a bit messy -- it'd actually be much easier to report *both* error message.

Do you think that's ok? If so, I can work on submitting a separate PR with this change after this one is landed. (I'll probably need to do some mild refactoring so we can reuse the "should we report a warning with this line?" logic, and I'd prefer to do that in a separate branch.)

File: test-data/unit/check-unreachable-code.test (hunk 7 above)
Comment (Michael0x2a) [reply test-data/unit/check-unreachable-code.test]: That could be handy, though possibly annoying to use for people with linters (since they don't understand mypy magic functions). We could work around this by adding the function to `mypy_extensions`? Or maybe that's overkill?

Another idea: once the "tag all error messages with a code" PR is landed, the user could maybe add a `# type: ignore` for just the "Statement is unreachable" warning?

**Comment by Michael0x2a** [comment]
@JukkaL -- Thanks for the CR! I updated the PR; this should be ready for a second look.

File: test-data/unit/check-unreachable-code.test (hunk 7 above)
Comment (Michael0x2a) [reply test-data/unit/check-unreachable-code.test]: Or another idea -- what if we instead modified mypy so that we *mandate* that any calls to `unreachable()` are actually unreachable?

We could also modify this function so that optionally take in any number of arguments: if the checker runs into this function, it'll run `reveal_type(...)` on the provided args before reporting an error that the statement was actually reachable. This would also let us supplant this "assert never" pattern (and providing a runtime implementation of the same logic in `mypy_extensions` or something would feel less like overkill).

**Comment by bluetech** [comment]
I tried this PR on some of our code. It found a bunch of useless `is None` checks (and a [typeshed bug](https://github.com/python/typeshed/pull/3093)), which is nice.

I was interested to check whether this would "break" the `assert_never` trick from #5818, which is like the opposite of this feature ("warn if this is *not* unreachable"), but the warning is not triggered in that case, which is good. I used this:

```py
from __future__ import annotations
from typing import Union, NoReturn


def assert_never(value: NoReturn) -> NoReturn:
    assert False, f'Unhandled value: {value} ({type(value).__name__})'


def f(x: Union[int, str]) -> None:
    if isinstance(x, int):
        pass
    elif isinstance(x, str):
        pass
    else:
        # Triggers "Statement is unreachable".
        # print('xx')
        # Doesn't trigger.
        assert_never(x)
```

File: test-data/unit/check-unreachable-code.test (hunk 7 above)
Comment (JukkaL) [reply test-data/unit/check-unreachable-code.test]: Yeah, that would be how I imagined it to work. Not sure about the arguments though. My idea would be to do something like this:

```py
X = Union[A, B]

def f(x: X) -> None:
    if isinstance(x, A):
        ...
    elif isinstance(x, B):
        ...
    else:
        unreachable()  # Error only if we forgot to handle some case
```

Maybe we could automatically recognize that pattern and print out the inferred type of `x` if we hit an error.

File: test-data/unit/check-unreachable-code.test (hunk 16 above)
Comment (JukkaL) [reply test-data/unit/check-unreachable-code.test]: The current behavior seems reasonable to me.

**Comment by ilevkivskyi** [comment]
Nice feature! Would be great to add docs for this.
